### PR TITLE
cranelift-module: expose trap information when defining functions

### PR DESCRIPTION
**Pull request by froydnj**
- [X] This has been discussed in issue #1184.
- [X] A short description of what this does, why it is needed: The current interface of `cranelift-module` requires consumers who want to be informed about traps to discover that information through `Module::Product`, which is backend-specific.  Since it's advantageous to manipulate this information in a backend-agnostic way, this patch changes `Module::define_function{,_bytes}` to return information about the traps contained in the function being defined.
- [X] This PR does not contain test cases.
- [x] A reviewer from the core maintainer team has been assigned for this PR.

### CODE REVIEW

File: cranelift/object/src/backend.rs [hunk 1]
```diff
@@ -188,7 +189,7 @@ impl Backend for ObjectBackend {
         ctx: &cranelift_codegen::Context,
         _namespace: &ModuleNamespace<Self>,
         code_size: u32,
-    ) -> ModuleResult<ObjectCompiledFunction> {
+    ) -> ModuleResult<(ObjectCompiledFunction, Option<&Vec<TrapSite>>)> {
```
Comment (bjorn3) [inline cranelift/object/src/backend.rs]: Please return `&[TrapSite]` instead.

**Comment by philipc** [comment]
My intent in https://github.com/bytecodealliance/wasmtime/issues/1184#issuecomment-590618409 was that the backends would no longer store traps at all. e.g. the caller would pass a `binemit::TrapSink` to `Module::define_function` (or a new variant of that), and `Module::define_function_bytes` would no longer need to know about traps. But it's not something that I strongly feel needs to change if you think it's better to keep storing them in the backend.

File: cranelift/object/src/backend.rs (hunk 1 above)
Comment (froydnj) [reply cranelift/object/src/backend.rs]: Good point, done.

**Comment by froydnj** [comment]
> My intent in [#1184 (comment)](https://github.com/bytecodealliance/wasmtime/issues/1184#issuecomment-590618409) was that the backends would no longer store traps at all. e.g. the caller would pass a `binemit::TrapSink` to `Module::define_function` (or a new variant of that), and `Module::define_function_bytes` would no longer need to know about traps. But it's not something that I strongly feel needs to change if you think it's better to keep storing them in the backend.

Ah, I didn't fully grok what you were saying there; thank you for the additional clarification!

I think passing in `binemit::TrapSink` would be slightly cleaner, but that implies that traps are somehow more important than relocs or stackmaps, and I'm not clear on why that'd be true.  I guess you can argue that the backend should be responsible for applying relocations on its own, and nobody seems to (currently) do anything with stackmaps anyway (at least in the context of `cranelift-module`)?

In the absence of strong feelings, I am inclined to leave the patch as-is.

**Comment by pchickey** [comment]
I agree that relocations are something we just give the backend.

We'll have to do something with stackmaps in `cranelift-module` eventually, the lack of pressure is because they are used for GC, which only wasmtime is going to implement, and wasmtime isnt using cranelift-module yet. There's a reasonable argument that whatever design we pick for traps, we should use for stack maps as well, for the sake of consistency. Passing in a sink for traps and stackmaps gives the consumer more control than returning a vector, so I'm in favor of that design, if we're going to revise.

I'm also ok with landing this as-is and revising it later, speaking as a downstream. Its not a huge difference to us either way, so I'll defer to bjorn and philipc.

**Comment by philipc** [comment]
Yeah this is ok as-is.

**Comment by pchickey** [comment]
Lets get @tschneidereit to add you as a collab so you can merge these when they're ready to go

**Comment by tschneidereit** [comment]
> Lets get @tschneidereit to add you as a collab so you can merge these when they're ready to go

Agreed!

**Pull request merged**